File: code/zig/14-heap_sort.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const allocator = std.heap.page_allocator;
const rand = std.rand;
const mem = std.mem;

/// Function to perform Heap Sort on an array.
///
/// # Parameters
///
/// - `arr`: The array to sort.
fn heapSort(arr: []usize) void {
    const n = arr.len;

    if (n <= 1) {
        return;
    }

    // Build a maxheap (rearrange array)
    var i = @divFloor(n, 2) - 1;
    while (i >= 0) : (i -= 1) {
        heapify(arr, n, i);
        if (i == 0) break;
    }

    // One by one extract elements from heap
    i = n - 1;
    while (i > 0) : (i -= 1) {
        // Move current root to end
        mem.swap(usize, &arr[0], &arr[i]);

        // Call max heapify on the reduced heap
        heapify(arr, i, 0);
    }
}

/// Function to heapify a subtree rooted with node `i`, which is an index in `arr`.
///
/// # Parameters
///
/// - `arr`: The array to heapify.
/// - `n`: The size of the heap.
/// - `i`: The index of the root node of the subtree.
fn heapify(arr: []usize, n: usize, i: usize) void {
    var largest: usize = i; // Initialize largest as root
    const left: usize = 2 * i + 1; // left child index
    const right: usize = 2 * i + 2; // right child index

    // If left child exists and is greater than root
    if (left < n and arr[left] > arr[largest]) {
        largest = left;
    }

    // If right child exists and is greater than largest so far
    if (right < n and arr[right] > arr[largest]) {
        largest = right;
    }

    // If largest is not root
    if (largest != i) {
        // Swap arr[i] with arr[largest]
        mem.swap(usize, &arr[i], &arr[largest]);

        // Recursively heapify the affected sub-tree
        heapify(arr, n, largest);
    }
}

/// Function to verify if an array is sorted.
///
/// # Parameters
///
/// - `arr`: The array to check.
fn isSorted(arr: []usize) bool {
    const n = arr.len;
    for (0..n - 1) |i| {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

/// Main function to demonstrate Insertion Sort and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 10_000, 20_000, 30_000, 40_000 };

    // Seed the random number generator
    const seed: u64 = @intCast(time.nanoTimestamp());
    var rng = rand.DefaultPrng.init(seed); // Xoshiro256 is good enough

    for (N_values) |N| {
        // Allocate memory for the array
        var arr = try allocator.alloc(usize, N);
        defer allocator.free(arr);

        // Fill the array with random integers
        for (0..N) |i| {
            const random_value: usize = rng.next() % N;
            arr[i] = random_value;
        }

        // Measure the start time
        const start = try Instant.now();

        // Perform Heap Sort
        heapSort(arr);

        // Measure the end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Verify that the array is sorted
        const is_sorted = isSorted(arr);

        // Print the result and time
        print("Insertion Sort with N = {d}\n", .{N});
        if (is_sorted) {
            print("Array is sorted.\n", .{});
        } else {
            print("Array is NOT sorted.\n", .{});
        }
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}

File: code/c/14-heap_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/**
 * Function to heapify a subtree rooted with node i which is an index in arr[].
 *
 * @param arr The array to heapify.
 * @param n The size of the heap.
 * @param i The index of the root node of the subtree.
 */
void heapify(int arr[], int n, int i) {
  int largest = i;        // Initialize largest as root
  int left = 2 * i + 1;   // left = 2*i + 1
  int right = 2 * i + 2;  // right = 2*i + 2

  // If left child exists and is greater than root
  if (left < n && arr[left] > arr[largest]) largest = left;

  // If right child exists and is greater than largest so far
  if (right < n && arr[right] > arr[largest]) largest = right;

  // If largest is not root
  if (largest != i) {
    // Swap arr[i] with arr[largest]
    int temp = arr[i];
    arr[i] = arr[largest];
    arr[largest] = temp;

    // Recursively heapify the affected sub-tree
    heapify(arr, n, largest);
  }
}

/**
 * Function to perform Heap Sort on an array.
 *
 * @param arr The array to sort.
 * @param n The size of the array.
 */
void heap_sort(int arr[], int n) {
  // Build a maxheap (rearrange array)
  for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);

  // One by one extract elements from heap
  for (int i = n - 1; i >= 0; i--) {
    // Move current root to end
    int temp = arr[0];
    arr[0] = arr[i];
    arr[i] = temp;

    // Call max heapify on the reduced heap
    heapify(arr, i, 0);
  }
}

/**
 * Main function to demonstrate Heap Sort and measure execution time.
 */
int main() {
  int N_values[] = {100000, 200000, 400000, 800000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    int *arr = (int *)malloc(N * sizeof(int));

    // Seed the random number generator
    srand(123);

    // Fill the array with random integers
    for (int i = 0; i < N; i++) {
      arr[i] = rand();
    }

    // Measure the start time
    clock_t start_time = clock();

    // Perform Heap Sort
    heap_sort(arr, N);

    // Measure the end time
    clock_t end_time = clock();

    // Calculate the elapsed time in seconds
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Verify that the array is sorted
    int is_sorted = 1;
    for (int i = 1; i < N; i++) {
      if (arr[i - 1] > arr[i]) {
        is_sorted = 0;
        break;
      }
    }

    // Print the result and time
    printf("Heap Sort with N = %d\n", N);
    if (is_sorted)
      printf("Array is sorted.\n");
    else
      printf("Array is NOT sorted.\n");
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free the allocated memory
    free(arr);
  }

  return 0;
}
